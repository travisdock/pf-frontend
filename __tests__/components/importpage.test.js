import React from 'react';
import { shallow } from 'enzyme';

import {
    nakedImportPage as NakedImportPage
} from '../../src/components/import_page';

// props for no redux
const userProp = {user_id: null}

// Mock fetches
function mockFetch(data) {
    return jest.fn().mockImplementation(() =>
      Promise.resolve({
        ok: true,
        json: () => data
      })
    );
  }

//   https://github.com/facebook/jest/issues/2157#issuecomment-279171856
function flushPromises() {
    return new Promise(resolve => setImmediate(resolve));
}

describe('<ImportPage /> with redux', () => {
    it('renders correctly', () => {
        const component = shallow(
            <NakedImportPage current_user={ userProp } />
        );
        expect(component).toMatchSnapshot();
    })
    it('handleSubmit function fires when button is pressed', async (done) => {
        const spy = jest.spyOn(NakedImportPage.prototype, "handleSubmit");
        window.alert = jest.fn()
        window.fetch = mockFetch({message: "error"})

        const component = shallow(
            <NakedImportPage current_user={ userProp } />
        );
        
        const form = component.find('form')
        await form.simulate('submit', { preventDefault: jest.fn(), target: {0: {files: {0: "file"}}} });

        expect(spy).toHaveBeenCalled();
        spy.mockRestore();
        done();
    });

    it('alerts with message from response', async (done) => {
        window.alert = jest.fn()
        window.fetch = mockFetch({message: "message from response"})

        const component = shallow(
            <NakedImportPage current_user={ userProp } />
        );

        const form = component.find('form')
        await form.simulate('submit', { preventDefault: jest.fn(), target: {0: {files: {0: "file"}}} });
        
        return flushPromises().then(() => {
            expect(window.alert).toHaveBeenCalledWith("message from response");
            done();
        });
    });

    it('fires action if import is successful', async (done) => {
        window.alert = jest.fn()
        window.fetch = mockFetch({message: "success message", categories: [{}, {}, {}]})
        const updateUser = jest.fn()

        const component = shallow(
            <NakedImportPage current_user={ userProp } updateUser={ updateUser } />
        );
        
        const form = component.find('form')
        await form.simulate('submit', { preventDefault: jest.fn(), target: {0: {files: {0: "file"}}} });
        
        return flushPromises().then(() => {
            expect(updateUser).toHaveBeenCalled();
            done();
        });
    });


});