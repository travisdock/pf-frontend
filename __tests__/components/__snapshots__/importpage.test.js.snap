// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImportPage /> with redux renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImportPage
    current_user={
      Object {
        "user_id": null,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="import_container"
      >
        <h2>
          Upload CSV
        </h2>
        <p
          align="center"
        >
          You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.
        </p>
        <div
          className="img_container"
        >
          <img
            alt="example format screenshot"
            src="/import_example.png"
          />
        </div>
        <form
          encType="multipart/form-data"
          onSubmit={[Function]}
        >
          <input
            id="fileupload"
            type="file"
          />
          <input
            id="submit"
            type="submit"
            value="submit"
          />
        </form>
      </div>,
      "className": "upper_third",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Upload CSV
          </h2>,
          <p
            align="center"
          >
            You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.
          </p>,
          <div
            className="img_container"
          >
            <img
              alt="example format screenshot"
              src="/import_example.png"
            />
          </div>,
          <form
            encType="multipart/form-data"
            onSubmit={[Function]}
          >
            <input
              id="fileupload"
              type="file"
            />
            <input
              id="submit"
              type="submit"
              value="submit"
            />
          </form>,
        ],
        "className": "import_container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Upload CSV",
          },
          "ref": null,
          "rendered": "Upload CSV",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "align": "center",
            "children": "You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.",
          },
          "ref": null,
          "rendered": "You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="example format screenshot"
              src="/import_example.png"
            />,
            "className": "img_container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "example format screenshot",
              "src": "/import_example.png",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                id="fileupload"
                type="file"
              />,
              <input
                id="submit"
                type="submit"
                value="submit"
              />,
            ],
            "encType": "multipart/form-data",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "fileupload",
                "type": "file",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "id": "submit",
                "type": "submit",
                "value": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": "form",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="import_container"
        >
          <h2>
            Upload CSV
          </h2>
          <p
            align="center"
          >
            You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.
          </p>
          <div
            className="img_container"
          >
            <img
              alt="example format screenshot"
              src="/import_example.png"
            />
          </div>
          <form
            encType="multipart/form-data"
            onSubmit={[Function]}
          >
            <input
              id="fileupload"
              type="file"
            />
            <input
              id="submit"
              type="submit"
              value="submit"
            />
          </form>
        </div>,
        "className": "upper_third",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h2>
              Upload CSV
            </h2>,
            <p
              align="center"
            >
              You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.
            </p>,
            <div
              className="img_container"
            >
              <img
                alt="example format screenshot"
                src="/import_example.png"
              />
            </div>,
            <form
              encType="multipart/form-data"
              onSubmit={[Function]}
            >
              <input
                id="fileupload"
                type="file"
              />
              <input
                id="submit"
                type="submit"
                value="submit"
              />
            </form>,
          ],
          "className": "import_container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Upload CSV",
            },
            "ref": null,
            "rendered": "Upload CSV",
            "type": "h2",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "align": "center",
              "children": "You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.",
            },
            "ref": null,
            "rendered": "You can upload csv files of previous financial information as long as it has Date, Category, Amount, and Note headers(case sensitive). It works best if the date category is formatted as a date in excel or google sheets before it is exported to csv. Any categories that do not already exist in your account will be created as a regular expense category.",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                alt="example format screenshot"
                src="/import_example.png"
              />,
              "className": "img_container",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "alt": "example format screenshot",
                "src": "/import_example.png",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  id="fileupload"
                  type="file"
                />,
                <input
                  id="submit"
                  type="submit"
                  value="submit"
                />,
              ],
              "encType": "multipart/form-data",
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "fileupload",
                  "type": "file",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "id": "submit",
                  "type": "submit",
                  "value": "submit",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "form",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
